Fixed #31185 -- Fixed detecting of unique fields in ForeignKey/ForeignObject checks when using Meta.constraints.

Valze opened the pull request:
Fixed #31185 -- Adjusted model check for UniqueConstraints without conditions
.
Changed _check_unique_target to take into account UniqueConstraints without conditions. Added test in invalid_models_tests, and changed code to take into account pre-existing method total_unique_constraints
Ticket number is [31185](https://code.djangoproject.com/ticket/31185)

hramezani commented:
@Valze, Also you need to add tests for this patch.
@carltongibson You missed a `j` in the PR title. `Adusted` -> `Adjusted`

carltongibson commented:
Good catch @hramezani!

Valze commented:
Thank you @hramezani and @carltongibson for your feedback!
I apologize for not checking out the commit requirements. I'll add tests, alter the commit message, and fix the flake8 failure

felixxm commented on django/db/models/fields/related.py:
```diff
@@ -528,6 +528,11 @@ def _check_unique_target(self):
             for ut in self.remote_field.model._meta.unique_together
         })
+        if self.remote_field.model._meta.constraints:
+            for constraint in self.remote_field.model._meta.constraints:
```
You should omit partial unique constraints, please use [`total_unique_constraints()`](https://github.com/django/django/blob/0352a44dd61c19bebf0c0b305dbbc3f710ff9945/django/db/models/options.py#L830-L840).

Valze commented:
> Hi @Valze. Thanks for this.
> 
>     * Initially, can you adjust the commit message [to match the guidelines](https://docs.djangoproject.com/en/3.0/internals/contributing/committing-code/#committing-guidelines). I've adjusted the PR title as a hint.
> 
>     * Can you also fix the flake8 failure.
> 
>     * Please do also add the test, using the example on the ticket. The related tests seem to be in `tests/invalid_models_tests/test_relative_fields.py`.

I'm a bit confused on why the test should be written inside the invalid_models_tests, and I think I may have written an incorrect test due to not understanding the rationale

I've fixed the flake8 failure. It's my first time working on a major codebase and I had no idea there was a python package purely to maintain PEP8. That is seriously cool.

I've taken into account what @felixxm said and I've used that nifty method.

Let me know what else I should change

hramezani commented on django/db/models/fields/related.py:
```diff
@@ -13,6 +13,7 @@
 from django.db.models.deletion import CASCADE, SET_DEFAULT, SET_NULL
 from django.db.models.query_utils import PathInfo
 from django.db.models.utils import make_model_tuple
+from django.db.models.constraints import UniqueConstraint
```
This is unused. Please remove it.

Valze replied:
Oh good catch. I forgot to remove the imports I made in my past tests. Thank you for the review. Fixing it now.

hramezani commented on tests/constraints/tests.py:
```diff
@@ -5,7 +5,7 @@
 from django.test import SimpleTestCase, TestCase, skipUnlessDBFeature
 
-from .models import ChildModel, Product, UniqueConstraintProduct
+from .models import ChildModel, Product, UniqueConstraintProduct, UniqueConstraintForeign
```
This is unused

Valze commented:
Fixing the flake8 errors before committing again. Still getting used to it

hramezani commented:
@Valze Thanks.
Please prepare your local development and run the `flake8` tests there for future contributions.
If your patch is ready please uncheck `Patch needs improvement` and `Needs tests` flags on the ticket.

felixxm commented:
That's not a proper test, we should add tests to the `RelativeFieldTests` for few scenarios:
- `ForeignObject()` to fields with `UniqueConstraint` (`fields.E310` is not raised),
- `ForeignObject()` to fields with partial `UniqueConstraint` (`fields.E310` is raised),
- `ForeignKey()` to a field with `UniqueConstraint` (`fields.E311` is not raised),
- `ForeignKey()` to a field with partial `UniqueConstraint` (`fields.E311` is raised).

Update also messages for `fields.E310` and `fields.E311`, they should mention `meta.constraints`.

Valze commented:
> That's not a proper test, we should add tests to the `RelativeFieldTests` for few scenarios:
> 
>     * `ForeignObject()` to fields with `UniqueConstraint` (`fields.E310` is not raised),
> 
>     * `ForeignObject()` to fields with partial `UniqueConstraint` (`fields.E310` is raised),
> 
>     * `ForeignKey()` to a field with `UniqueConstraint` (`fields.E311` is not raised),
> 
>     * `ForeignKey()` to a field with partial `UniqueConstraint` (`fields.E311` is raised).
> 
> 
> Update also messages for `fields.E310` and `fields.E311`, they should mention `meta.constraints`.

I understand how i should write the tests now. Should I still work on the patch or should I leave it to you since you assigned it to yourself?

felixxm commented:
> I understand how i should write the tests now. Should I still work on the patch or should I leave it to you since you assigned it to yourself?

Yes please add tests.

felixxm commented:
@Valze Thanks :+1: I pushed edits. Welcome aboard :boat:

Valze commented:
You don't even know how happy I am. I'm such a huge fan of django that contributing, as small of a bugfix as this is, is an absolute honor.

Looking forward to contributing more
